src/viur/core/modules/user.py

```diff
@@ -821,6 +853,140 @@ def transaction(user_key, idx):
+            raise errors.PreconditionFailed("No 'otp_app_secret' is provided")
+        if not (cuser := current.user.get()):
+            raise errors.Unauthorized()
+
+        def transaction(user_key):
+            if not (user := db.Get(user_key)):
+                raise errors.NotFound()
+            user["otp_app_secret"] = otp_app_secret
+            db.Put(user)
+
+        db.RunInTransaction(transaction, cuser["key"])
+
+    @classmethod
+    def generate_otp_app_secret_uri(cls, otp_app_secret) -> str:
+        """
+        :return an otp uri like otpauth://totp/Example:alice@google.com?secret=ABCDEFGH1234&issuer=Example
+        """
+        if not (cuser := current.user.get()):
+            raise errors.Unauthorized()
+        if not (issuer := conf["viur.otp.issuer"]):
+            logging.warning(
+                f"""conf["viur.otp.issuer"] is None we replace the issuer by conf["viur.instance.project_id"]""")
+            issuer = conf["viur.instance.project_id"]
+
+        return pyotp.TOTP(otp_app_secret).provisioning_uri(name=cuser["name"], issuer_name=issuer)
+
+    @classmethod
+    def generate_otp_app_secret(cls) -> str:
+        """
+        Generate a new OTP Secret
+        :return an otp
+        """
+        return pyotp.random_base32()
+
+    @classmethod
+    def verify_otp(cls, otp: str | int, secret: str) -> bool:
+        return pyotp.TOTP(secret).verify(otp)
+
+    @exposed
+    def start(self):
+        return self._user_module.render.edit(TimeBasedOTP.OtpSkel(),
+                                             action_name=self.ACTION_NAME,
+                                             action_url=self.action_url,
+                                             tpl=self.second_factor_login_template)
+
+    @exposed
+    @force_ssl
+    @skey
+    def authenticator_otp(self, **kwargs):
+        """
+        We verify the otp here with the secret we stored before.
+        """
+
+        session = current.session.get()
+        user_key = db.Key(self._user_module.kindName, session["possible_user_key"])
+
+        if not (user := db.Get(user_key)):
+            raise errors.NotFound()
+
+        skel = TimeBasedOTP.OtpSkel()
+        if not skel.fromClient(kwargs):
+            raise errors.PreconditionFailed()
+        otp_token = str(skel["otptoken"]).zfill(6)
+
+        if AuthenticatorOTP.verify_otp(otp=otp_token, secret=user["otp_app_secret"]):
+            return self._user_module.secondFactorSucceeded(self, user_key)
+        else:
+            skel = TimeBasedOTP.OtpSkel()
+            skel.errors = [ReadFromClientError(ReadFromClientErrorSeverity.Invalid, "Wrong OTP Token")]
+            return self._user_module.render.edit(skel,
+                                                 action_name=self.ACTION_NAME,
+                                                 action_url=self.action_url,
+                                                 tpl=self.second_factor_login_template)
```

Comment:
can you simplify, and please re-use the OtpSkel instance.
```suggestion

        skel.errors = [ReadFromClientError(ReadFromClientErrorSeverity.Invalid, "Wrong OTP Token")]
        return self._user_module.render.edit(
            skel,
            action_name=self.ACTION_NAME,
            action_url=self.action_url,
            tpl=self.second_factor_login_template
        )
```